### PR TITLE
Light position params

### PR DESCRIPTION
**Pull request by karimnaaji**
#598

### CODE REVIEW

File: core/src/scene/light.h [hunk 1]
```diff
@@ -27,6 +27,15 @@ enum class LightOrigin {
     world
 };
 
+static std::string lightOriginString(LightOrigin origin) {
```
Comment (tallytalwar) [inline core/src/scene/light.h]: You can make this a member function of the `LightOrigin` enum class. Something like `toString`.

Comment (blair1618) [inline core/src/scene/light.h]: Alas, I really wish c++ typed enums allowed this - Java enums do and it's very very convenient.

**Comment by blair1618** [comment]
Ok, we do need to parse these units correctly, this approach should work.. however I do feel that our treatment of value-unit parameters is in need of some love. We do unit parsing in too many ways in too many places (where n > 1 is too many). There are some details that I'll comment on in-line, but this approach will do for now with the understanding that we can (should) revisit units more holistically in the near future.

File: core/src/scene/styleParam.h [hunk 2]
```diff
@@ -60,6 +60,28 @@ constexpr size_t StyleParamKeySize = static_cast<size_t>(StyleParamKey::NUM_ELEM
...
+        case Unit::meter: return "meter";
+        case Unit::seconds: return "seconds";
+        default: return "undefined";
+    }
+}
+
+template <typename T, int size>
+struct UnitVec {
+    T value = T(0.0);
+    Unit units[size];
```
Comment (blair1618) [inline core/src/scene/styleParam.h]: I wonder if we could do `Unit units[sizeof(T)/sizeof(T[0])];`

**Comment by karimnaaji** [comment]
Yep that's what I noticed too; those unit conversion really need a refactoring after we all get that working.

File: core/src/scene/styleParam.h (hunk 2 above)
Comment (karimnaaji) [inline core/src/scene/styleParam.h]: Will give that a try.

**Comment by blair1618** [comment]
*poke* @karimnaaji this still has conflicts

**Comment by karimnaaji** [comment]
Yes, rebasing this in a few.

File: core/src/scene/styleParam.h (hunk 2 above)
Comment (karimnaaji) [inline core/src/scene/styleParam.h]: Unfortunately, this does not work.. :/

Comment (blair1618) [inline core/src/scene/styleParam.h]: *facepalm* my bad, it makes no sense to index into a type name haha. `sizeof(value)/sizeof(value[0])` should work though. Then we can remove the `size` template parameter.

Comment (karimnaaji) [inline core/src/scene/styleParam.h]: Even in that case it wound't be a `constexpr` right?

Comment (blair1618) [inline core/src/scene/styleParam.h]: It actually is! At least in C++14. The values are all definitely known at compile-time, I think arithmetic might not have been allowed until C++14.

**Comment by blair1618** [comment]
Lots of warnings about `unitString` as an unused function. It's a natural thing to have, but maybe we can comment it for now :\

**Pull request merged**